File: app/src/main/java/com/arifian/training/wisatasemarang/HomeFragment.java
package com.arifian.training.wisatasemarang;


import android.content.Intent;
import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ProgressBar;

import com.arifian.training.wisatasemarang.adapters.WisataAdapter;
import com.arifian.training.wisatasemarang.databinding.FragmentHomeBinding;
import com.arifian.training.wisatasemarang.models.Wisata;
import com.arifian.training.wisatasemarang.models.remote.SimpleRetrofitCallback;
import com.arifian.training.wisatasemarang.models.remote.responses.WisataResponse;

import org.parceler.Parcels;

import java.util.ArrayList;
import java.util.List;

import static com.arifian.training.wisatasemarang.DetailWisataActivity.KEY_WISATA;


/**
 * A simple {@link Fragment} subclass.
 */
public class HomeFragment extends Fragment {
    FragmentHomeBinding mBinding;

    List<Wisata> wisataArrayList = new ArrayList<>();
    WisataAdapter adapter;

    public HomeFragment() {
        // Required empty public constructor
    }

    public static HomeFragment newInstance() {

        Bundle args = new Bundle();

        HomeFragment fragment = new HomeFragment();
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        mBinding = FragmentHomeBinding.inflate(inflater, container, false);

        adapter = new WisataAdapter(wisataArrayList, wisata -> {
            Intent intent = new Intent(getActivity(), DetailWisataActivity.class);
            intent.putExtra(KEY_WISATA, Parcels.wrap(wisata));
            startActivity(intent);
        });
        mBinding.recyclerView.setAdapter(adapter);

        getWisata();

        return mBinding.getRoot();
    }

    private void getWisata() {
        ProgressBar progressBar = new ProgressBar(getActivity(), null, android.R.attr.progressBarStyleSmall);

        progressBar.setVisibility(View.VISIBLE);
        WisataApplication.get(getActivity())
                .getService(getActivity())
                .getWisata()
                .enqueue(new SimpleRetrofitCallback<WisataResponse>(getActivity()) {
                    @Override
                    public void onSuccess(WisataResponse response) {
                        wisataArrayList.addAll(response.getWisata());
                        adapter.notifyDataSetChanged();
                    }
                });
    }
}

File: app/src/main/java/com/arifian/training/wisatasemarangkotlin/HomeFragment.kt
package com.arifian.training.wisatasemarangkotlin


import android.content.Intent
import android.os.Bundle
import android.support.v4.app.Fragment
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ProgressBar
import com.arifian.training.wisatasemarangkotlin.Utils.Constants.Companion.KEY_WISATA
import com.arifian.training.wisatasemarangkotlin.adapters.WisataAdapter
import com.arifian.training.wisatasemarangkotlin.databinding.FragmentHomeBinding
import com.arifian.training.wisatasemarangkotlin.models.Wisata
import com.arifian.training.wisatasemarangkotlin.models.remote.SimpleRetrofitCallback
import com.arifian.training.wisatasemarangkotlin.models.remote.responses.WisataResponse
import org.parceler.Parcels
import java.util.*


/**
 * A simple [Fragment] subclass.
 */
class HomeFragment : Fragment() {
    internal lateinit var mBinding: FragmentHomeBinding

    internal var wisataArrayList: MutableList<Wisata> = ArrayList()
    internal lateinit var adapter: WisataAdapter

    override fun onCreateView(inflater: LayoutInflater?, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        // Inflate the layout for this fragment
        mBinding = FragmentHomeBinding.inflate(inflater!!, container, false)

        adapter = WisataAdapter(wisataArrayList, object: WisataAdapter.OnWisataClickListener{
            override fun onItemClick(wisata: Wisata) {
                val intent = Intent(activity, DetailWisataActivity::class.java)
                intent.putExtra(KEY_WISATA, Parcels.wrap(wisata))
                startActivity(intent)
            }
        })

        mBinding.recyclerView.adapter = adapter

        getWisata()

        return mBinding.root
    }

    private fun getWisata() {
        val progressBar = ProgressBar(activity, null, android.R.attr.progressBarStyleSmall)

        progressBar.visibility = View.VISIBLE
        WisataApplication.get(activity)
                .getService(activity)
                .wisata
                .enqueue(object : SimpleRetrofitCallback<WisataResponse>(activity) {
                    override fun onSuccess(response: WisataResponse) {
                        wisataArrayList.addAll(response.wisata!!)
                        adapter.swapData(wisataArrayList)
                    }
                })
    }

    companion object {

        fun newInstance(): HomeFragment {

            val args = Bundle()

            val fragment = HomeFragment()
            fragment.arguments = args
            return fragment
        }
    }
}// Required empty public constructor
